events/janus_gelfevh.c

```diff
@@ -0,0 +1,589 @@
+			JANUS_GELFEVH_ERROR_MISSING_ELEMENT, JANUS_GELFEVH_ERROR_INVALID_ELEMENT);
+		if(error_code != 0)
+			goto plugin_response;
+		/* Parameters we can change */
+		const char *req_events = NULL, *req_backend = NULL, *req_port = NULL;
+		int req_compress = -1, req_compression = -1;
+		/* Events */
+		if(json_object_get(request, "events"))
+			req_events = json_string_value(json_object_get(request, "events"));
+		/* Compression */
+		if (json_object_get(request, "compress"))
+			req_compress = json_is_true(json_object_get(request, "compress"));
+		if (json_object_get(request, "compression"))
+			req_compress = json_integer_value(json_object_get(request, "compression"));
+		/* Backend stuff */
+		if(json_object_get(request, "backend"))
+			req_backend = json_string_value(json_object_get(request, "backend"));
+		if(json_object_get(request, "port"))
+			req_port = json_string_value(json_object_get(request, "port"));
+		if (json_object_get(request, "max_message_len"))
+			max_gelf_msg_len = json_integer_value(json_object_get(request, "max_message_len"));
+		if (strcasecmp(json_string_value(json_object_get(request, "protocol")), "udp") == 0){
+			transport = UDP;
+		}
+		if(!req_backend || !req_port) {
+			/* Invalid backend address or port */
+			error_code = JANUS_GELFEVH_ERROR_INVALID_ELEMENT;
+			g_snprintf(error_cause, sizeof(error_cause), "Invalid backend URI '%s:%s'", req_backend, req_port);
+			goto plugin_response;
+		}
+		/* If we got here, we can enforce */
+		janus_mutex_lock(&evh_mutex);
+		if(req_events)
+			janus_events_edit_events_mask(req_events, &janus_gelfevh.events_mask);
+		if (req_compress > -1)
+			compress = req_compress ? TRUE : FALSE;
+		if (req_compression > -1 && req_compression < 10)
+			compression = req_compression;
+		if(req_backend && req_port) {
+			g_free(backend);
+			g_free(port);
+			backend = g_strdup(req_backend);
+			port = g_strdup(req_port);
+		}
+		janus_mutex_unlock(&evh_mutex);
+	} else {
+		JANUS_LOG(LOG_VERB, "Unknown request '%s'\n", request_text);
+		error_code = JANUS_GELFEVH_ERROR_INVALID_REQUEST;
+		g_snprintf(error_cause, 512, "Unknown request '%s'", request_text);
+	}
+
+plugin_response:
+		{
+			json_t *response = json_object();
+			if(error_code == 0) {
+				/* Return a success */
+				json_object_set_new(response, "result", json_integer(200));
+			} else {
+				/* Prepare JSON error event */
+				json_object_set_new(response, "error_code", json_integer(error_code));
+				json_object_set_new(response, "error", json_string(error_cause));
+			}
+			return response;
+		}
+}
+
+/* Thread to handle incoming events */
+static void *janus_gelfevh_handler(void *data) {
+	JANUS_LOG(LOG_VERB, "Joining GelfEventHandler handler thread\n");
+	json_t *event = NULL;
+
+	while(g_atomic_int_get(&initialized) && !g_atomic_int_get(&stopping)) {
+		event = g_async_queue_pop(events);
+		if(event == NULL)
+			continue;
+		if(event == &exit_event)
+			break;
+
+		/* Handle event */
+		while(TRUE) {
```

Comment:
`while(TRUE)` is a bad idea, there should be checks on whether things are shutting down. Check the other handlers to see how loops are handled there.

Reply:
this is the event handling thread that I took from simple:
https://github.com/meetecho/janus-gateway/blob/master/events/janus_sampleevh.c#L461
and i see it also in new ws event handler:
https://github.com/meetecho/janus-gateway/blob/master/events/janus_wsevh.c#L543

Reply:
there is a check before that, if it is not stopping and initialized

> while(g_atomic_int_get(&initialized) && !g_atomic_int_get(&stopping)) {

Reply:
Fair point, I should actually check the code I wrote before complaining :slightly_smiling_face: